Parse PluralRules from CLDR

kostya9 opened the pull request:
Fully generated file with rules is here https://gist.github.com/kostya9/467b404cddfbaae2fa63b5e6c6bfb584

I used the specification here https://www.unicode.org/reports/tr35/tr35-numbers.html#Language_Plural_Rules to guide me.

I implemented a source generator that
1. Parses pluralrules.xml (taken from CLDR latest http://cldr.unicode.org/index/downloads) into a intermediate representations.
2. Generates c# code for each rule there
3. The MessageFormatter fallbacks to metadata if it can't find a rule for such locale in it's personal Pluralizer dictionary

Feel free to start review either from tests, or from the source generator entrypoint PluralLanguagesGenerator.cs.

For variables, I implemented only 
- n | absolute value of the source number.
- i | integer digits of n.
- v | number of visible fraction digits in n, with trailing zeros. (partially - only figured out how to determine whether a number is fractional or not. In most languages the value v=0 or v!=0 is used, excluded the languages with other values of v from metadata)
Locales with other variables were excluded

For right side, the generator understands both ranges (11..14) and individual numbers (15).
For left side, the generator understands both plain variables(i), and module operations (i % 10).

Contributes to https://github.com/jeffijoe/messageformat.net/issues/22

UPD:
Added support for all variables except exponents

kostya9 committed:
Add source generator. Parse simple rules with absolute value and visible digits number

kostya9 committed:
Fix assembly name. Refactor rule parsing to a class

kostya9 committed:
Add 'and' rule parsing

kostya9 committed:
Add modulo support

kostya9 committed:
Add support for range operators in the right part

kostya9 committed:
Add source generation for a single rule

kostya9 committed:
Add code generation

kostya9 committed:
Make locale functions public

kostya9 committed:
Extract helpers for indents

kostya9 committed:
Move equals logic to left operand

kostya9 committed:
Add test for full source code

kostya9 commented:
Hey @jeffijoe could you please update the .NET SDK to 5.0.* in github actions?

kostya9 commented on src/Jeffijoe.MessageFormat/Jeffijoe.MessageFormat.csproj:
```diff
@@ -23,4 +23,17 @@
+  </ItemGroup>
+
+  <PropertyGroup>
+    <PluralLanguagesMetadataExcludeLocales>si da is mk ceb fil tl lv prg bs hr sh sr fr dsb hsb lt</PluralLanguagesMetadataExcludeLocales>
```
These languages contain constructs that this generator cannot parse yet

jeffijoe replied:
`da` is Danish? Damn, sucks to be me 😂

kostya9 replied:
Eventually, support for these should be there, but I figured that it would be even harder to review with more changes 😅

kostya9 commented on src/Jeffijoe.MessageFormat/Jeffijoe.MessageFormat.csproj:
```diff
@@ -10,11 +10,11 @@
   	<LangVersion>latest</LangVersion>
   	<Nullable>enable</Nullable>
-  	<TargetFramework>netstandard1.1</TargetFramework>
+  	<TargetFramework>netstandard2.0</TargetFramework>
```
Was forced to  update to netstandard2.0 to consume the source generator. Feel free to evaluate whether this is worth it or not

jeffijoe replied:
What are we losing by doing this?

NightOwl888 replied:
The [.NET Standard](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) page documents this, but in layman's terms:

- .NET Framework 4.5 through 4.6 (4.5 and 4.5.1 have reached EoL).
- .NET Core 1.x (which has reached EoL).
- Mono 4.5 through 5.3
- All flavors of Xamarin and UWP would have to be upgraded slightly (probably not an issue for most)

For now keeping a single .NET Standard 2.0 target is the best option for supporting both .NET Framework and .NET Core without adding all of the dependencies that .NET Standard 1.x had.

You may also wish to consider the tradeoffs of supporting earlier versions of .NET Framework and improved API support for .NET Standard 2.1 which you could gain at the price of multi-targeting and keeping conditional sections to gracefully degrade for .NET Framework, both of which add complexity. But IMO, abandoning .NET Standard 1.x at this stage is a good decision either way.

jeffijoe replied:
So TL;DR is that upgrading to 2.0 is what we want to do either way? 😄

NightOwl888 replied:
Maybe. If multi-targeting .NET Framework 4.5 and .NET Standard 2.1, there isn't a huge gap that .NET Standard 2.0 fills (basically .NET Core 2.x and some older mobile platforms). It depends on how high your tolerance for setting up a more complicated build and for dealing with conditional compilation sections is :wink:.

jeffijoe replied:
I don't mind removing support for EoL targets.

kostya9 commented on src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs:
```diff
@@ -139,7 +139,10 @@ internal string Pluralize(string locale, ParsedArguments arguments, double n, do
             if (this.Pluralizers.TryGetValue(locale, out pluralizer) == false)
             {
-                pluralizer = this.Pluralizers["en"];
+                if(PluralRulesMetadata.TryGetRuleByLocale(locale, out pluralizer) == false)
```
This is absolutely open to discussion, currently, the metadata is used whenever we can't find a user-defined locale with such a name

jeffijoe replied:
I think that makes sense. I would imagine eventually we remove support for custom pluralizers once we support all of the official locale rules.

kostya9 commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -0,0 +1,251 @@
+
+            var numberSpan = ConsumeCharacters(numbersCount);
+            
+            var number = int.Parse(new string(numberSpan.ToArray()));
```
Didn't find span-based overload in netstandard 2.0 :(

jeffijoe replied:
`numberSpan.ToString()` might be more efficient than `ToArray()` + `new string()`?

kostya9 replied:
Didn't know you could do that, thanks

jeffijoe commented:
> Hey @jeffijoe could you please update the .NET SDK to 5.0.* in github actions?

Isn't that controlled in the workflow file in the branch?

kostya9 committed:
Bump SDK version

kostya9 committed:
Update test project version

kostya9 commented:
Thanks, that worked! Wasn't sure of how the security model works with forks, but seems that I can change SDKs freely.

jeffijoe commented:
This is amazing! I'm going to need some time to play with it and review, but wow! 🤩

NightOwl888 commented:
Rather than passing "Locale" as a string in .NET, it would be more intuitive (and more compatible) to pass in a `CultureInfo` instance. Of course, this would create some gaps that you would need to account for, but probably less so in .NET 5 which now uses ICU by default.

Also, it is quite non-intuitive to attach `CultureInfo` (or Locale) to an object instance in .NET by passing it through a constructor. It would be better if it could be passed to the method that does the formatting. This would also have the advantage that a single formatter could work with multiple threads at the same time. Ideally, there would be 2 overloads of `FormatMessage()`, one that accepts a `CultureInfo` and one that uses the `CultureInfo` from the current thread.

I am not saying you have to make these changes, but this is the way Microsoft would do it if they add support for message formatting.

kostya9 committed:
Remove not used code

kostya9 commented:
Accepting CultureInfo in method argument sounds reasonable, but would not satisfy my use-case. My situation is that multiple users are being served by one server. These users can have different languages configured for their personal account. Basically, I have a language code, and need to format a string for that language code.

I imagine that going from a language code to a CultureInfo may be a bit awkward.

kostya9 commented:
I agree with your point that it would have been easier to pass the locale to the formatting method, though. Probably, that sounds like a separate discussion thread, and the parsing logic for CLDR rules will be the same regardless of whether the locale-specific pluralization data is needed in the constructor or in the formatting method.

NightOwl888 commented:
> Accepting CultureInfo in method argument sounds reasonable, but would not satisfy my use-case. My situation is that multiple users are being served by one server. These users can have different languages configured for their personal account. Basically, I have a language code, and need to format a string for that language code.

There isn't a lot of pluralization data and doesn't need to be updated at runtime. It could be made as a default set of pluralization rules that is cached.

But the fact that you are serving multiple users per server is the point. If this is a webserver, each user can pass their culture in the URL or somewhere else in the request envelope and it can then be made "the" culture of the current thread.

```c#
CultureInfo.CurrentCulture = CultureInfo.GetCultureInfo(userCulture);
```

This is done at the beginning of each request, so when using the `MessageFormatter.FormatMessage()` you don't necessarily have to pass the culture at that point because it is already set on the thread for the user. In fact, if pulling the data out of a read-only cache (or it looks like you may have it all codified all of the rules at that point), you could potentially have a singleton `MessageFormatter` that serves all users on all threads.

> I imagine that going from a language code to a CultureInfo may be a bit awkward.

> I agree with your point that it would have been easier to pass the locale to the formatting method, though. Probably, that sounds like a separate discussion thread, and the parsing logic for CLDR rules will be the same regardless of whether the locale-specific pluralization data is needed in the constructor or in the formatting method.

I haven't analyzed this in detail to see how big the change would be. But most users will be using the built-in `CultureInfo` object because it is attached to the current thread already. In fact, not using the one from the current thread would probably seem odd to most .NET developers because culture is context-sensitive by default in many .NET features.

I believe you could use `CultureInfo.TwoLetterISOLanguageName` to convert most cultures to locale, but there will probably be some gaps that need to be mapped from .NET to the CLDR name. It would be better for end users if closing those gaps was built-in rather than leaving it up to each one to roll their own solution.

Agreed it probably doesn't change the parsing of the file, but maybe change how it is dealt with at the top level after it is parsed.  I just noticed it was being passed in to a constructor and wanted to point out that isn't the .NET way because .NET is already aware of its current culture at the top level. It would be best if `MessageFormatter` were aware of that culture and provide a way to override it.

NightOwl888 commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/InvalidCharacterException.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Jeffijoe.MessageFormat.MetadataGenerator.Plural.Parsing
+{
+    public class InvalidCharacterException : ArgumentException
```
.NET has an exception for failing to parse: [FormatException](https://docs.microsoft.com/en-us/dotnet/api/system.formatexception?view=net-5.0).

kostya9 replied:
Forgot about that one for some reason, thanks

NightOwl888 commented on src/Jeffijoe.MessageFormat.Tests/Jeffijoe.MessageFormat.Tests.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFramework>net5.0</TargetFramework>
```
I would suggest multi-targeting the tests. This adds enough complexity that it may differ on .NET Framework vs .NET Core vs .NET 5.

jeffijoe replied:
Good idea, is this easy to do?

NightOwl888 replied:
Yes, just change the test projects from using

```xml
<TargetFramework>net5.0</TargetFramework>
```

to

```xml
<TargetFrameworks>net5.0;netcoreapp3.1;net48</TargetFrameworks>
```

> **TIP:** Make sure you pay attention to the `s` in the name or you might be scratching your head for awhile.

In Visual Studio 2019, it will then show you all of the tests on each of the target frameworks and you can run them all or select the ones you want to run.

![image](https://user-images.githubusercontent.com/1538288/114055679-30cd4f80-98bb-11eb-8744-5b422160a356.png)


At the command line, it is a bit more complicated because AFAIK you have to run a `dotnet test` on each target framework as a separate command.

I believe it will work just to install the .NET 5.0 SDK (or at least you can always use the .NET 5 SDK as the entry point and all of the options it supports, but you may have to install other SDKs such as .NET Core 3.1 before installing .NET 5.0).

### Working Example

My setup is more complicated than yours because I test on multiple operating systems and use build assets to transfer the same binaries from the build agent to the test agents (you could just run `dotnet test` which would build from source on each OS, but I prefer to test the binaries I ship). I am also using Azure DevOps rather than GitHub Actions, but you should be able to set it up similarly there, too if you want to.

In my case, I do `dotnet publish` (in combination with properly setting the `<IsPublishable>` property to `true` for test projects and `false` for all of the others) to make the tests runnable on each OS and I specifically set up a build agent to test 1 target framework. But you are welcome to analyze the templates [here](https://github.com/NightOwl888/Morfologik.Stemming/blob/master/azure-pipelines.yml) and [here](https://github.com/NightOwl888/Morfologik.Stemming/tree/master/build/azure-templates) and use what you need (from anywhere in the project) to get it working.

NightOwl888 commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/PluralLanguagesGenerator.cs:
```diff
@@ -0,0 +1,59 @@
+namespace Jeffijoe.MessageFormat.MetadataGenerator
+{
+    [Generator]
+    public class PluralLanguagesGenerator : ISourceGenerator
```
Brilliant use of a source code generator!

kostya9 committed:
Use proper null character for EOF, use appropriate FormatException for formatting errors

kostya9 committed:
Remove duplicate logic for IsEnd check, as PeekCurrentChar does absolutely the same check

kostya9 commented:
The more I think about it, the more your point about `CultureInfo` makes sense. I tried to parse a couple of locales via `CultureInfo.GetCultureInfo`, and they just work. Will need to research if it works for all languages that I care about.

WIll not do any changes in this PR, but the SourceGenerator should work if somewhere in the future the `CultureInfo` approach will be introduced. The partial method `PluralRulesMetadata.TryGetRuleByLocale` will just be used differently `PluralRulesMetadata.TryGetRuleByLocale(CultureInfo.TwoLetterISOLanguageName, out var formatter);`

jeffijoe commented:
Sorry this is taking a while, I have a lot going on these days 😪

kostya9 commented:
No worries, the diff is quite large too, so take your time

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -0,0 +1,253 @@
+                else
+                {
+                    throw new InvalidCharacterException(character);
+                }
```
It's definitely negligible, but we can defer reading the next character in case we have an early return.

```suggestion
                var character = ConsumeChar();
                // This is where the samples start, we don't care about any of those.
                if (character == '@')
                {
                    return conditions.ToArray();
                }

                // We expect the next token to be "or"
                var characterNext = ConsumeChar();
                if (character == 'o' && characterNext == 'r')
                {
                    continue;
                }
                
                throw new InvalidCharacterException(character);
```

kostya9 replied:
I like the change! Going with it

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -0,0 +1,253 @@
+        {
+            if (_position + count > _ruleText.Length)
+            {
+                var characters = _ruleText.AsSpan(_position, _ruleText.Length - _position);
```
I appreciate your use of `Span`! 🙌

jeffijoe commented on src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/RuleSourceGeneratorTests.cs:
```diff
@@ -0,0 +1,366 @@
+        [Fact]
+        public void CanGenerateRuleForFractionNumberEquals()
+        {
+            var generator = new RuleGenerator(new PluralRule(new[] { "en" }, new[]
```
I would not be opposed to using the parser here instead of constructing the AST manually. I know that is what I did with the message formatting tests but I've found its more trouble than its worth; I've been writing a programming language parser after I wrote the initial version of MF.NET and have found that to be a nicer experience.

If you want to keep it this way then that's cool too, just want you to know that its an option!

kostya9 replied:
Eh, put too much effort to remove them now. But I won't mind if someone will refactor these in the future

jeffijoe commented:
@kostya9 Working through the PR, looking great so far!

* Are there any tests for one of the generated rules?
* Should there not be an entry for `zero`? Or was adding that to the built-in `en` pluralizer a mistake on my part?
* > v | number of visible fraction digits in n, with trailing zeros. (partially - only figured out how to determine whether a number       is fractional or not. In most languages the value v=0 or v!=0 is used, excluded the languages with other values of v from metadata)
  
  I think [`Round`](https://docs.microsoft.com/en-us/dotnet/api/system.math.round?view=net-5.0#System_Math_Round_System_Double_System_Int32_) + string format would do the trick?

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -0,0 +1,253 @@
+                if (PeekCurrentChar == 'a')
+                {
+                    var andWord = ConsumeCharacters(3);
+                    ReadOnlySpan<char> andWordExpected = stackalloc char[3]
+                    {
+                        'a',
+                        'n',
+                        'd'
+                    };
```
Nice use of `stackalloc`! Let's move this to the top of the method though, so we don't allocate on the stack for every iteration.

I am curious though, is this the most performant and efficient way to check for tokens? Asking in general because I'm working on another parser in a different project, so wondering if you have some insight here that I am not aware of 👀 

**EDIT:** I experimented a bit with a basic benchmark of `stackalloc` vs `AsSpan` on a `const string`. It looks like `AsSpan` is slightly faster but naturally allocates the original string once.

This is the benchmark and results if you are interested:

```
|     Method |       N |           Mean |         Error |       StdDev | Ratio | Rank |        Gen 0 | Gen 1 | Gen 2 |     Allocated |
|----------- |-------- |---------------:|--------------:|-------------:|------:|-----:|-------------:|------:|------:|--------------:|
|  Substring |    1000 |     2,591.9 us |   2,119.39 us |    116.17 us |  1.00 |    3 |    1414.0625 |     - |     - |    11837801 B |
| InlineSpan |    1000 |       495.5 us |      72.86 us |      3.99 us |  0.19 |    2 |            - |     - |     - |             - |
|     AsSpan |    1000 |       466.9 us |     171.25 us |      9.39 us |  0.18 |    1 |            - |     - |     - |             - |
|            |         |                |               |              |       |      |              |       |       |               |
|  Substring | 1000000 | 2,531,425.7 us | 866,693.02 us | 47,506.37 us |  1.00 |    3 | 1425000.0000 |     - |     - | 11922191088 B |
| InlineSpan | 1000000 |   486,633.6 us |   7,166.70 us |    392.83 us |  0.19 |    2 |            - |     - |     - |         288 B |
|     AsSpan | 1000000 |   473,793.4 us | 141,400.46 us |  7,750.64 us |  0.19 |    1 |            - |     - |     - |         288 B |

```

```csharp
    [RankColumn]
    [MemoryDiagnoser]
    public class SpanBenchmark
    {
        static readonly string[] words = new[]
        {
            "lorem", "ipsum", "dolor", "sit", "amet", "consectetuer",
            "adipiscing", "elit", "sed", "diam", "nonummy", "nibh", "euismod",
            "tincidunt", "ut", "laoreet", "dolore", "magna", "aliquam", "erat"
        };

        // borrowed from greg (https://stackoverflow.com/questions/4286487/is-there-any-lorem-ipsum-generator-in-c)
        static IEnumerable<string> LoremIpsum(Random random, int minWords, int maxWords, int minSentences, int maxSentences, int numLines)
        {
            var line = new StringBuilder();
            for (int l = 0; l < numLines; l++)
            {
                line.Clear();
                var numSentences = random.Next(maxSentences - minSentences) + minSentences + 1;
                for (int s = 0; s < numSentences; s++)
                {
                    var numWords = random.Next(maxWords - minWords) + minWords + 1;
                    line.Append(words[random.Next(words.Length)]);
                    for (int w = 1; w < numWords; w++)
                    {
                        line.Append(" ");
                        line.Append(words[random.Next(words.Length)]);
                    }
                    line.Append(". ");
                }
                yield return line.ToString();
            }
        }

        string[] lines;

        [Params(1000, 1_000_000)]
        public int N;

        [GlobalSetup]
        public void GlobalSetup()
        {
            lines = LoremIpsum(new Random(), 6, 8, 2, 3, 1_000_000).ToArray();
        }

        public static int CountOccurrencesSub(string val)
        {
            if (string.IsNullOrEmpty(val))
            { return 0; }

            int count = 0;

            for (int x = 0; x <= val.Length - searchFor.Length; x++)
            {
                if (val.Substring(x, searchFor.Length) == searchFor)
                { count++; }
            }

            return count;
        }

        public static int CountOccurrences(string val)
        {
            if (string.IsNullOrEmpty(val))
            { return 0; }
            int count = 0;

            ReadOnlySpan<char> vSpan = val.AsSpan();
            ReadOnlySpan<char> searchSpan = stackalloc char[]
            {
                'l',
                'o',
                'r',
                'e',
                'm',
                ' ',
                'i',
                'p',
                's',
                'u',
                'm',
                ' ',
                'd',
                'o',
                'l',
                'o',
                'r',
                ' ',
                's',
                'i',
                't',
                ' ',
                'a',
                'm',
                'e',
                't',
                ' ',
                'c',
                'o',
                'n',
                's',
                'e',
                'c',
                't',
                'e',
                't',
                'u',
                'e',
                'r'
            };

            for (int x = 0; x <= vSpan.Length - searchSpan.Length; x++)
            {
                if (vSpan.Slice(x, searchSpan.Length).SequenceEqual(searchSpan))
                { count++; }
            }

            return count;
        }
        
        const string searchFor = "lorem ipsum dolor sit amet consectetuer";
        
        public static int CountOccurrencesAsSpan(string val)
        {
            if (string.IsNullOrEmpty(val))
            { return 0; }

            int count = 0;

            ReadOnlySpan<char> vSpan = val.AsSpan();
            ReadOnlySpan<char> searchSpan = searchFor.AsSpan();

            for (int x = 0; x <= vSpan.Length - searchSpan.Length; x++)
            {
                if (vSpan.Slice(x, searchSpan.Length).SequenceEqual(searchSpan))
                { count++; }
            }

            return count;
        }
        
        [Benchmark(Baseline = true)]
        public int Substring()
        {
            int occurences = 0;
            for (var i = 0; i < N; i++)
                occurences += CountOccurrencesSub(lines[i]);
            return occurences;
        }

        [Benchmark]
        public int InlineSpan()
        {
            int occurences = 0;
            for (var i = 0; i < N; i++)
                occurences += CountOccurrences(lines[i]);
            return occurences;
        }
        
        [Benchmark]
        public int AsSpan()
        {
            int occurences = 0;
            for (var i = 0; i < N; i++)
                occurences += CountOccurrencesAsSpan(lines[i]);
            return occurences;
        }
```

kostya9 replied:
I was aiming for this trick (see Optimized C# on the right panel) https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0ATEBqAPgAQCYBGAWACh8AGAAn2IDoARKAQwHcBLAOwHMBuChXwBmOoRoBhGhQDeFGorpj6SGgCUYrDAHluAGwCeAZQAOrbgB4wAC1ZQAfDQCq3CKYAunALacAXjAYABQAlDQAvE7cMOw0tvYA2gC6NLIA5KxpaDRpwFk5YGkAvoLkSjJlSqJ0xGqa2npGZhaWwIYeME46nj7+gaERUTE0bR3JqVQIxNmThDMIIiUURUA==
but forgot to check what actually happens. Seems that the trick works for `ReadOnlySpan<byte>` only.

kostya9 replied:
Used AsSpan (stackalloc - a case of overengineering from me 😄 ), moved out of the loop

jeffijoe replied:
How about a source generator for our source generator? 🤔 😂

kostya9 replied:
There is one 😉 https://github.com/ufcpp/StringLiteralGenerator

Found it here https://github.com/dotnet/roslyn/issues/48850

jeffijoe replied:
Oh my lord this is amazing. We gotta use it, if for anything do it for the meme! 😂

jeffijoe replied:
<img width="487" alt="image" src="https://user-images.githubusercontent.com/2505474/114068028-73673b80-986b-11eb-8e27-0cdc56bf87e9.png">

pointnet commented:
Latvian uses [v = 2](https://github.com/unicode-org/cldr-json/blob/master/cldr-json/cldr-core/supplemental/plurals.json#L472)
French uses [e = 0](https://github.com/unicode-org/cldr-json/blob/master/cldr-json/cldr-core/supplemental/plurals.json#L217)

Only way to have something relevant is to parse strings like [here](https://github.com/pointnet/alrev-intl/blob/develop/packages/Alrev.Intl.PluralRules/PluralRulesContext.cs#L74)

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/LeftOperand.cs:
```diff
@@ -0,0 +1,54 @@
+﻿namespace Jeffijoe.MessageFormat.MetadataGenerator.Plural.Parsing
+{
+    public interface ILeftOperand
```
If you want, you can put these and the other AST nodes into their own files in a `AST` sub-directory

kostya9 replied:
Sounds good, done

NightOwl888 commented on src/Jeffijoe.MessageFormat/Jeffijoe.MessageFormat.csproj:
```diff
@@ -10,11 +10,11 @@
 
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
-    <DocumentationFile>bin/Release/netstandard1.1/Jeffijoe.MessageFormat.xml</DocumentationFile>
+    <DocumentationFile>bin/Release/netstandard2.0/Jeffijoe.MessageFormat.xml</DocumentationFile>
```
This can be parameterized, which might make maintenance simpler:

```xml
<DocumentationFile>bin/$(Configuration)/$(TargetFramework)/$(AssemblyName).xml</DocumentationFile>
```

kostya9 replied:
Thanks, used this

kostya9 commented:
@jeffijoe 
> Are there any tests for one of the generated rules?

Wasn't sure how to test that, I can't possibly test all of the languages. I can write a couple of test cases for languages I speak (en-ru-uk), what do you think about that?

> Should there not be an entry for zero? Or was adding that to the built-in en pluralizer a mistake on my part?

There seems to be no zero-rule there
https://github.com/unicode-org/cldr/blob/master/common/supplemental/plurals.xml#L33

> I think Round + string format would do the trick?

Yeah, that should work. Didn't do that yet because was thinking: maybe there is another way without allocating a string when calculating the variables?

kostya9 committed:
Address feedback

Move AST to a separate folder, move classes into separate files.
Use parametrized expression for documentation file.
Apply simplification for ParseRuleContent.
Remove 'and' calculation out of loop.
Simplify number construction.

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -198,6 +199,8 @@ private IRightOperand[] ParseRightOperand()
 
         private OrCondition ParseOrCondition()
         {
+            var andWordSpan = "and".AsSpan();
```
I don't know if this actually makes a difference but you can make it a `private const string` at the class level.

kostya9 replied:
Looks the same to me
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0ATEBqAPgAQCYBGAWACgL8BmAAiNoGUAHAQwDsBnCgbwtoG1mUAJYA3VgBcYtSF0n1iABlocMtALy0ARGu0BufoOHipMkewUBhWGYCC7DAFVOMKxHmNJo9gHMAFACURgJ85IIRtBJQtJxs7JqqjgB0dpwsHEGG4ZEC+ADssfHJADIwfpIAFtkRAL4UIfR0FgoAIhBeAK4AZt0A6iJV7p7eFgHBOaGNEdGxnQC2iUrZ04Ld0LT+syJL+rQ7ADy0ysv72NgTuVOTVwKcC7TYWjYw9o4ubh6ckl4+4zVXeqUG4RApzeYAgRAxomCTSfaWWgvN7OVwASXYABsLDBfmMgo0wrdZnEOIldI5tKl0vEsqs8oVSexSuVfFVIbRoTcaAi2h1JD1+oNKhjsexcaM/ASbkSriSHlplg0QWsNltWDEdoq9odjkpTiJzpcrrLbncHk8kbZpA5UTBRTi8VLAhy6srbmD7hDGkDakA

jeffijoe replied:
Nice! Thanks for verifying!

jeffijoe commented:
> Wasn't sure how to test that, I can't possibly test all of the languages. I can write a couple of test cases for languages I speak (en-ru-uk), what do you think about that?

Yeah that sounds good! You may be able to use the compiler to run the generated class?

> Yeah, that should work. Didn't do that yet because was thinking: maybe there is another way without allocating a string when calculating the variables?

Could you give me an example of what you mean here?

kostya9 committed:
Add locale-specific tests to verify plural rules are there

kostya9 commented:
> Yeah that sounds good! You may be able to use the compiler to run the generated class?

The generated rules are in tests assembly already, because MessageFormat assembly already contains the generated code when referenced. Added tests. Made sure they work by commenting out the line which uses generated metadata and ensuring the tests fail (except EN, of course).

> Could you give me an example of what you mean here?

I mean that string.Format creates(allocates) a new string, was hesitant to use it for that reason. I wonder if there is an easy enough mathematical method (divide the initial number by something. multiply by something) to calculate these variables. If there is no other reasonable way - well, that's better than nothing :)

kostya9 committed:
Minor rename - indicate that test tests Plural rules

jeffijoe commented:
@kostya9 oh I completely misunderstood what the `v` was actually for! Can you give me an example of a rule where the basic `v!=/==0` does not apply?

~EDIT: Thinking about it more, it really doesn't make sense that an input number will have trailing zeros because number types don't even encode that information. Unless we are saying that you may format a number-like `string` in which case we have to parse the string.~

EDIT2: I completely forgot about the `decimal` data type which does in fact encode that information!

We currently convert to `double` here:

https://github.com/jeffijoe/messageformat.net/blob/0ac587c8f418768509ea8378d0dba4313a60f640/src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs#L103

If we change from `double` to `decimal` then we should be able to use [this approach](https://stackoverflow.com/a/13493771/561545):

```csharp
int count = BitConverter.GetBytes(decimal.GetBits(value)[3])[2];
```

kostya9 commented:
@jeffijoe Finally found some time to work on this :)

I'm exploring adding string manipulation for fractional numbers to support all the rules properly (trailing zeroes and stuff).
Could you please add some context on what is 'offset' here ?https://github.com/jeffijoe/messageformat.net/blob/0ac587c8f418768509ea8378d0dba4313a60f640/src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs#L97

jeffijoe commented:
> Could you please add some context on what is 'offset' here ?

Yes, see this: https://messageformat.github.io/messageformat/guide/#plural-offset

kostya9 committed:
Support all rules except exponent via introducing PluralContext

kostya9 commented:
Added support for all rules except exponent (you cannot express exponent in any way except passing string with exponent directly)
See the new version of generated file here https://gist.github.com/kostya9/c95fa4395811c0d3d24e7654b563c322

Thanks @pointnet , did something like you proposed

kostya9 commented on src/Jeffijoe.MessageFormat/Jeffijoe.MessageFormat.csproj:
```diff
@@ -10,11 +10,11 @@
   	<LangVersion>latest</LangVersion>
   	<Nullable>enable</Nullable>
-  	<TargetFramework>netstandard1.1</TargetFramework>
+  	<TargetFrameworks>netstandard2.0;net5.0</TargetFrameworks>
```
Added net5.0 to target frameworks to allocate less when parsing PluralContext on projects that support it

kostya9 committed:
Remove double parsing

kostya9 committed:
Fix culture assumption

kostya9 committed:
Add special case for double in context

kostya9 committed:
Search for dot character, not string

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/RuleParser.cs:
```diff
@@ -0,0 +1,256 @@
+
+            var numberSpan = ConsumeCharacters(numbersCount);
+            
+            var number = int.Parse(numberSpan.ToString());
```
Would adding the .NET 5 condition here help avoid the `ToString`?

kostya9 replied:
Unfortunately no, to the best of my knowledge. AFAIK source generators must target netstandard2.0 (I suppose, that has something to do with Visual Studio being a NETFramework application )

jeffijoe commented on src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs:
```diff
@@ -0,0 +1,100 @@
+                PluralRule rule = _rules[ruleIdx];
+                foreach (var locale in rule.Locales)
+                {
+                    WriteLine($"{{\"{locale}\", (ContextPluralizer) Rule{ruleIdx}}},");
```
Is the cast here necessary?

kostya9 replied:
Nope, it's useless, removed that

kostya9 committed:
Remove useless cast in generated code

kostya9 committed:
Remove bunch of ToArrays in SourceGenerator

jeffijoe commented:
Amazing work @kostya9, and shoutout to @pointnet and @NightOwl888 for the input! Much appreciated!

This looks good to merge, unless you have some final cleanup you want to do first?

kostya9 committed:
Make plural metadata internal

kostya9 commented:
Thanks! Made a final pass through the code - made the plural metadata internal. Everything else looks good to me, feel free to merge

jeffijoe commented:
Would you happen to have R# or Rider? Maybe running a code cleanup before merge would be good hygiene 😄

kostya9 commented:
My Rider instance generates an insane diff when I try to do that )

I suppose it can't infer the code style properly, it's clueless without an explicit editorconfig

<img src="https://user-images.githubusercontent.com/3115913/116111157-0bb04d80-a6bf-11eb-9c7f-603603f9f52c.png" height="500">

jeffijoe commented:
I think there's a `DotSettings` file in the repo, granted it was generated many years ago when I was just using ReSharper. All good then. I just saw some places where the code was not formatted but it's fine. :)
